### DISCUSSION TITLE
ugly `_bound` suffixes methods everywhere

### DISCUSSION THREAD

**Discussion by jimy-byerley** [PyO3/pyo3#4089] [General]
Hello everyone
I wanted to ask what are these `_bound` methods everywhere in pyo3 0.21 ?!
I can understand the desire to make a dedicated struct for bound references, but yet it is redundant with `&'py` references and duplicates the whole API. Actually there is now 3 types of references to the same python object `Py<T>`, `&'py T`, `Bound<'py, T>`.
Are we supposed to replace all our pyo3 user code with only functions suffixed by `_bound` ? this is quite ugly and verbose. 

Why not simply replacing all `&'py T` by `Bound<'py, T>` in previous existing API ?
Why not keeping `&'py T` ?

**Comment by davidhewitt** [comment]
The goal is to replace all `&'py T` by `Bound<'py, T>`. Given the scale of the change we wanted to do this in a way where users can do this incrementally.

The `_bound` suffix is a temporary addition to allow this incremental migration. I completely agree with you that it's verbose, the goal is for PyO3 0.23 to remove the need for that suffix once users have had time to migrate.

https://pyo3.rs/v0.21.2/migration#from-020-to-021
https://polar.sh/davidhewitt/posts/replacing-pyo3-api-pt1

**Reply by jimy-byerley** [reply]
Okay, fair enough :sweat_smile: Now that I read your blog I realize it is a very deep change in progress in pyo3 !

**Reply by jimy-byerley** [reply]
Yet thanks for creating all this !
